### PR TITLE
Server repository unit tests

### PR DESCRIPTION
**Pull request by alalamav**
Adds unit tests for `OutlineServerRepository`.

### CODE REVIEW

File: src/www/app/outline_server.spec.ts [hunk 1]
```diff
@@ -92,6 +92,216 @@ describe('OutlineServerRepository', () => {
...
+        resolve();
+      });
+      eventQueue.startPublishing();
+    });
+  });
+
+  it('add throws on invalid access keys', () => {
+    const repo = new OutlineServerRepository(
+        getFakeServerFactory(), new EventQueue(), new InMemoryStorage());
+    expect(() => repo.add('ss://invalid')).toThrow();
```
Comment (mpmcroy) [inline src/www/app/outline_server.spec.ts:132]: Do we have concrete error types that we can check for?

File: src/www/app/outline_server.spec.ts [hunk 2]
```diff
@@ -92,6 +92,216 @@ describe('OutlineServerRepository', () => {
...
+    expect(servers[1].name).toEqual(CONFIG_1.name);
+  });
+
+  it('add emits ServerAdded event', async () => {
+    const eventQueue = new EventQueue();
+    const repo =
+        new OutlineServerRepository(getFakeServerFactory(), eventQueue, new InMemoryStorage());
+    const accessKey = shadowsocksConfigToAccessKey(CONFIG_0);
+    repo.add(accessKey);
+    await new Promise<void>(resolve => {
```
Comment (mpmcroy) [inline src/www/app/outline_server.spec.ts:118]: I don't think this catches the failure of the `ServerAdded` event not getting fired. Does the test fail if the promise doesn't resolve in time?

How about instead we await on `eventQueue.startPublishing()` (seems like it's synchronous) and `fail()` if we don't receive the event?

File: src/www/app/outline_server.spec.ts [hunk 3]
```diff
@@ -92,6 +92,216 @@ describe('OutlineServerRepository', () => {
...
+  it('forget emits ServerForgotten events', async () => {
+    const eventQueue = new EventQueue();
+    eventQueue.subscribe(ServerAdded, () => {});  // Silence dropped event warnings.
+    const repo =
+        new OutlineServerRepository(getFakeServerFactory(), eventQueue, new InMemoryStorage());
+    repo.add(shadowsocksConfigToAccessKey(CONFIG_0));
+    repo.add(shadowsocksConfigToAccessKey(CONFIG_1));
+    const forgottenServerId = repo.getAll()[0].id;
+    repo.forget(forgottenServerId);
+    await new Promise<void>(resolve => {
```
Comment (mpmcroy) [inline src/www/app/outline_server.spec.ts:233]: Same as above for the other emit event tests.

File: src/www/app/outline_server.spec.ts [hunk 4]
```diff
@@ -92,6 +92,216 @@ describe('OutlineServerRepository', () => {
...
+    expect(() => repo.validateAccessKey('ss://invalid')).toThrow();
+    // IPv6 host.
+    expect(() => repo.validateAccessKey(shadowsocksConfigToAccessKey({
+      host: '0:0:0:0:0:0:0:1',
+      port: 443,
+      password: 'test',
+      method: 'chacha20-ietf-poly1305'
+    }))).toThrow();
+    // Unsupported ciphers.
+    expect(
```
Comment (mpmcroy) [inline src/www/app/outline_server.spec.ts:296]: Same as above for error types.

File: src/www/app/outline_server.spec.ts (hunk 1 above)
Comment (alalamav) [reply src/www/app/outline_server.spec.ts]: Done in b67381b. I hhad tried `toThrow`, which compares error instances; now using `toThrowError`, which takes an error constructor.

File: src/www/app/outline_server.spec.ts (hunk 2 above)
Comment (alalamav) [reply src/www/app/outline_server.spec.ts]: jasmine fails tests that don't complete after 3s, so this works as expected. I made these tests synchronous per your suggestion in 54492e2.

File: src/www/app/outline_server.spec.ts (hunk 3 above)
Comment (alalamav) [reply src/www/app/outline_server.spec.ts]: Done in 54492e2.

File: src/www/app/outline_server.spec.ts (hunk 4 above)
Comment (alalamav) [reply src/www/app/outline_server.spec.ts:296]: Done in b67381b.

**Review by mpmcroy — APPROVED** [review]

**Pull request merged**